ssz_serialization/merkleization.nim

```diff
@@ -120,23 +120,82 @@ func digest*(a: openArray[byte]): Digest {.noinit.} =
         h.update(a)
         h.finish()
 
-func digest(a, b: openArray[byte]): Digest =
-  result = computeDigest:
-    trs "DIGESTING ARRAYS ", toHex(a), " ", toHex(b)
-    trs toHex(a)
-    trs toHex(b)
-
-    h.update a
-    h.update b
-  trs "HASH RESULT ", result
+{.compile: "hashtree.c".}
+
+when defined(gcc):
+  when defined(arm):
+    {.compile: "../vendor/hashtree/src/sha256_armv8_neon_x1.S".}
+    {.compile: "../vendor/hashtree/src/sha256_armv8_neon_x4.S".}
+
+  elif defined(amd64):
+    {.compile: "../vendor/hashtree/src/sha256_avx_x1.S".}
+    {.compile: "../vendor/hashtree/src/sha256_avx_x4.S".}
+    {.compile: "../vendor/hashtree/src/sha256_avx_x8.S".}
+    {.compile: "../vendor/hashtree/src/sha256_avx_x16.S".}
+    {.compile: "../vendor/hashtree/src/sha256_shani.S".}
+    {.compile: "../vendor/hashtree/src/sha256_sse_x1.S".}
+
+type Hasher = proc(output: pointer, input: pointer, count: uint64) {.cdecl, noSideEffect, raises: [Defect].}
+
+func digest64x(output: pointer, input: pointer, count: uint64) {.cdecl, noSideEffect, raises: [Defect].} =
+  cast[ptr Digest](output)[] =
+     digest(cast[ptr UncheckedArray[byte]](input).toOpenArray(0, 63))
+  discard
+
+func best_sha256_implementation(): Hasher {.importc.}
+
+let digest64* = block:
```

Comment:
digest64 is a really bad name. Maybe `digest2x` or `digest_2way`?

Reply:
well .. it's 64 bytes which is the significant requirement here .. ie unlike the other digest functions, it only works for blocks of that length.

Reply:
just call it `digestBlockSize` :)